Add unit test for scancode-evaluate.py and improve analysis of scancode results

harmut01 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).
    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
Fixes bug in `scancode-evaluate.py` where files were being incorrectly labelled as non-permissively licensed. Also added unit testing files: 
- `scancode_evaluate_test.py` 
- folder for stub files 'scancode_test/' 
- stub files: 4 JSON, 4 source files 

#### Impact of changes <!-- Optional -->
Impacts CI testing
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
None 
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
None 
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@hugueskamba @ARMmbed/mbed-os-core
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

harrisonmutai-arm committed:
Add unit test for script that checks copyright and license notice

Add unit test and dependent stubs for testing of scancode-evaluate.py.
license_check takes a JSON as argument and checks for source missing
copyright or license notices.

Each JSON represents a separate test case, and stubs are added for test case
3 and 4 where license_check looks for an spdx notice in the source file.

hugueskamba reviewed (commented):
Please replace all tabs in `tools/test/travis-ci/scancode_evaluate_test.py` with spaces.

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,106 @@
+			@outputs -1 if any error in file licenses found
+		"""
+		expected_result = -1
+		test_json = ROOT + "/scancode_test/scancode_test_1.json"
```
```suggestion
```

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,106 @@
+		expected_result = -1
+		test_json = ROOT + "/scancode_test/scancode_test_1.json"
+
+		# pass json path to test function
```
```suggestion
```

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,106 @@
+		# pass json path to test function
+		result = license_check(ROOT, test_json)
+
+		self.assertEqual(expected_result, result)
```
```suggestion
		self.assertEqual(
		    -1,
		    license_check(ROOT, ROOT + "/scancode_test/scancode_test_1.json"
		 )
```

Similar changes in `test_scancode_case_2()`.

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,106 @@
+		test_json = ROOT + "/scancode_test/scancode_test_1.json"
+
+		# pass json path to test function
+		result = license_check(ROOT, test_json)
```
```suggestion
```

hugueskamba commented on tools/test/travis-ci/scancode_test/scancode_test_1.json:
```diff
@@ -0,0 +1,7 @@
+        "tool_name": "scancode test fail"
+      }
+    ]
+  }
```
Create this file on the fly.
Also the indentation is off.
```suggestion
{
  "headers": [
    {
      "tool_name": "scancode test fail"
    }
  ]
}
```

rwalton-arm commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -110,6 +103,29 @@ def license_check(directory_name, file):
+        """
+        found_spdx = False
+        # iterate through licenses, stop once permissive license has been found
+        for i in range(len(license_offender['file']['licenses'])):
```
The `for` statement will automatically iterate over elements of a sequence. So, instead of explicitly calling `range(len(...` you can just have something like:

```
for license in license_offender['file']['licenses']:
    if license['category'] == 'Permissive':
    ...
```

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,106 @@
+				else:
+					file.write(invalid_header_1)
+
+		expected_result = 7
```
Is this correct? You expect 7 but the method docstring says it should be 5 and the scancode_test_3.json has 5 files. Can you explain the 7? What are the 7 errors?

harmut01 replied:
Apologies, I forgot to update the docstring! My understanding is that we want to be able to address the errors in the file individually.  For this reason two of the stubs contain a non-permissive license and no SPDX to test this. In total 4 errors are attributed to these two files (2 non-permissive, 2 SPDX missing).

harmut01 replied:
I've updated the docstring to reflect this - I hope it makes sense...

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,106 @@
+		os.remove(stub_path + "test5.h")
+
+	def test_scancode_case_4(self):
+		""" Test Case 4 -- license header permissive and non-permissive 'license' [FP]
```
What is `[FP]`?

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,106 @@
+
+		result = license_check(ROOT, test_json)
+
+		self.assertEqual(expected_result, result, "Non-Permissive Header False Positive")
```
How can you expect it to pass if `scancode_test_4.json` has one license missing `spdx` in `matched_rule.identifier`? Am I missing something?

harrisonmutai-arm committed:
Add unit test for script that checks copyright and license notice

Add unit test and dependent stubs for testing of scancode-evaluate.py.
license_check takes a JSON as argument and checks for source missing
copyright or license notices.

Each JSON represents a separate test case, and stubs are added for test
case 3 and 4 where license_check looks for an spdx notice in the source file.

harrisonmutai-arm committed:
Add unit test for script that checks copyright and license notice

Add unit test and dependent stubs for testing of scancode-evaluate.py.
license_check takes a JSON as argument and checks for source missing
copyright or license notices.

Each JSON represents a separate test case, and stubs are added for test
case 3 and 4 where license_check looks for an spdx notice in the source
file.

rwalton-arm commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,92 @@
+            @outputs -1 if any error in file licenses found
+        """
+        # generate scancode_test_1.json
+        with open(ROOT + "/scancode_test/scancode_test_1.json", 'w') as f:
```
Use `os.path` or `pathlib` instead of raw strings for paths, this ensures cross-platform compatibility.

rwalton-arm commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,92 @@
+            json.dump({"headers": {"tool name": "scancode test fail"}}, f)
+
+        self.assertEqual(-1, license_check(ROOT, ROOT + "/scancode_test/scancode_test_1.json"))
+        os.remove(ROOT + "/scancode_test/scancode_test_1.json")
```
If an exception is raised and the program exits before we get here, the files will not be removed. Consider using `tempfile.TemporaryDirectory` as a context manager to ensure clean-up happens in an exception-safe way.

rwalton-arm commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,92 @@
+import json
+from unittest import TestCase
+
+# TODO: fix scancode to match python naming conventROOTi
```
Please fix "conventROOTi" typo.

rwalton-arm commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,92 @@
+)
+
+# path to stub files
+stub_path = ROOT + "/scancode_test/"
```
```suggestion
stub_path = os.path.join(ROOT, "scancode_test")
```

rwalton-arm commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,92 @@
+        with open(ROOT + "/scancode_test/scancode_test_1.json", 'w') as f:
+            json.dump({"headers": {"tool name": "scancode test fail"}}, f)
+
+        self.assertEqual(-1, license_check(ROOT, ROOT + "/scancode_test/scancode_test_1.json"))
```
Can we define some named constants for the return codes so it's easier to understand what "-1" or "7" means?

rwalton-arm commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -0,0 +1,92 @@
+        self.assertEqual(-1, license_check(ROOT, ROOT + "/scancode_test/scancode_test_1.json"))
+        os.remove(ROOT + "/scancode_test/scancode_test_1.json")
+
+    def test_scancode_case_2(self):
```
```suggestion
    def test_scancode_no_errors_in_license_headers(self):
```
`test_scancode_case_2` is not a useful test case name. Same comment for other test cases.
